src/actions/wapeSale/purchaseDomains.ts

```diff
@@ -0,0 +1,121 @@
+import * as ethers from "ethers";
+import { getSaleStatus } from ".";
+import { WapeSale } from "../../contracts/types";
+import { Claim, SaleStatus, Mintlist, Maybe } from "../../types";
+
+const abi = ["function masterCopy() external view returns (address)"];
+
+const errorCheck = async (condition: boolean, errorMessage: string) => {
+  if (condition) {
+    throw errorMessage;
+  }
+};
+
+const generateAccessList = (
+  userAddress: string,
+  gnosisSafeProxyAddress: string,
+  gnosisSafeImplAddress: string
+) => {
+  return [
+    {
+      address: gnosisSafeProxyAddress,
+      storageKeys: [
+        "0x0000000000000000000000000000000000000000000000000000000000000000",
+      ],
+    },
+    {
+      address: gnosisSafeImplAddress,
+      storageKeys: [
+        "0x0000000000000000000000000000000000000000000000000000000000000000",
+      ],
+    },
+    {
+      address: userAddress,
+      storageKeys: [],
+    },
+  ];
+};
+
+export const purchaseDomains = async (
+  count: ethers.BigNumber,
+  signer: ethers.Signer,
+  contract: WapeSale,
+  mintlist: Mintlist
+): Promise<ethers.ContractTransaction> => {
+  const status = await getSaleStatus(contract);
+
+  errorCheck(
+    status === SaleStatus.NotStarted,
+    "Cannot purchase a domain when sale has not started or has ended"
+  );
+
+  errorCheck(status === SaleStatus.Ended, "Sale has already ended");
```

Comment:
I wonder if this is accounting for technical debt. Did `SaleStatus.NotStarted` at some point indicate either start or end, but then an upgrade added an actual `SaleStatus.Ended`.

I'd caution against combining the two.  Rather it might be preferable to update the error text of line 47 to something like `Cannot purchase domain, sale status is "NotStarted"`

Reply:
Edit: Since we're trying to avoid making changes at this stage, maybe just leave this as is.